subprojects/core/src/main/java/org/gradle/api/internal/tasks/timeout/DefaultTaskTimeoutHandler.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.timeout;
+
+import java.time.Duration;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+public class DefaultTaskTimeoutHandler implements TaskTimeoutHandler {
+    private final ScheduledExecutorService executor;
+
+    public DefaultTaskTimeoutHandler(ScheduledExecutorService executor) {
+        this.executor = executor;
+    }
+
+    @Override
+    public TaskTimeout start(Thread taskExecutionThread, Duration timeout) {
+        InterruptOnTimeout interrupter = new InterruptOnTimeout(taskExecutionThread, timeout.toMillis());
+        ScheduledFuture<?> periodicCheck = executor.scheduleAtFixedRate(interrupter, 0, 50, TimeUnit.MILLISECONDS);
```

Comment:
Would it make sense to pass an `initialDelay` of `timeout.toMillis()` here?

Reply:
Yes it would.